lib/db/observed.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import (
+	"time"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+func (db *Lowlevel) AddOrUpdatePendingDevice(device protocol.DeviceID, name, address string) error {
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	od := ObservedDevice{
+		Time:    time.Now().Round(time.Second),
+		Name:    name,
+		Address: address,
+	}
+	bs, err := od.Marshal()
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending device entry: %v", err)
+	}
+	return err
+}
+
+// PendingDevices drops any invalid entries from the database after a
+// warning log message, as a side-effect.  That's the only possible
+// "repair" measure and appropriate for the importance of pending
+// entries.  They will come back soon if still relevant.
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		var od ObservedDevice
+		deviceID := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
```

Comment:
I'd prefer fixing `DeviceIDFromBytes()`, but IIRC we agreed to keep that out of this PR.

Noted on my todo list so it won't get lost.

Reply:
Sounds good. If an issue helps you keeping track of things, by all means open one - otherwise there's no need, as there's no actual (known) bug related to it.

What I am asking for here is to please remove all the device ID length checks in this PR then.

Reply:
I thought it would be nice to replace them 1:1 by checks for a future `error` return value of `DeviceIDFromBytes()` once implemented, therefore the FIXME.

But I guess I can just as well look at fixing that function next in a separate PR branch and then merge into this one if you're going to squash anyway in the end.

Reply:
I rather not merge something into master with todo/fixme comments.

That would be a perfect candidate for a separate PR, meaning separate commit in master in the end. I.e. now remove the length checking, then we merge it, then file another PR with the error return (you can already start now working on that on top of this branch if you want to). In the end I don't think you should spend too much time thinking about git history (I am guilty of that too at times) - it's a valuable tool, but it's not the product.

Reply:
You're right about Git history in general, but it's a very valuable information resource IMHO so I think treating it with some care will pay out sometime.

I'm reluctant to "remove" the length check because without detecting errors from *or before* `DeviceIDFromBytes()` the reaction (`goto deleteKey`) is gone as well. So any later fix will change *what* this function does to handle the error, not only *how* it detects the condition. The other PR (slowly working on it) should be limited to the latter in all possible places to keep the changes isolated, should anyone need to bisect this later.

Reply:
I can change the wording and make it a TODO like in some other places. It's not really a fix**ME**, but rather a "fix that *other place* which is unfortunate to rely on here".  But don't worry I don't intend to leave it there for long. And anyways, the current solution does work flawlessly and is easier to debug than a silent panic in a REST handler goroutine.

Reply:
I am tired of the argument on device id length checks - I thought we already implicitly agreed to disagree on the benefit of those. And my point after that disagreement was, that I'd prefer if those checks were in place everywhere or nowhere. Now they are in place her, but not anywhere else. You apparently feel these checks are too important for such considerations. So in the end, leave it as is/how you want it, I am definitely not going to jeopardize this good PR on such a unimportant side-show. (And I don't care about the exact wording at all.)

Reply:
I also thought we had agreed on a separate PR. Just didn't want to hold up this one. But anyway, here goes: #6714.

Will revisit this once that change is approved and merged.

Reply:
I pulled in master again after #6714 got merged, so this is now properly implemented as of b67132f1d5b2ebdb7ecdf1377b6fcbdc261de880.